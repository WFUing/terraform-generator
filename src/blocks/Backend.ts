import { Argument, Attribute } from '../arguments';
import { BlockArgs } from '../utils';
import { Block } from '.';

/**
 * @category Block
 */
export class Backend extends Block {

  readonly type: string;

  /**
   * Construct backend.
   *
   * Refer to Terraform documentation on what can be put as type & arguments.
   *
   * @param type type
   * @param args arguments
   */
  constructor(type: string, args: BlockArgs) {
    super('backend', [type], args, undefined, true);

    this.type = type;
  }

  override asArgument(): Argument {
    return new Argument(JSON.stringify(this.type));
  }

  override attr(_name: string): Attribute {
    throw new Error('Inaccessible method.');
  }

}
