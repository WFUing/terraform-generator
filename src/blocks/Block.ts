import { TerraformVersion, Attribute, Argument } from '..';
import TerraformGeneratorUtils, { ArgumentsToStringFn } from '../TerraformGeneratorUtils';

export default abstract class Block {

  readonly blockType: string;
  readonly blockNames: string[];
  private readonly arguments: Record<string, any>;
  private readonly customArgumentToString: ArgumentsToStringFn;

  /**
   * Construct block.
   * 
   * @param type type
   * @param names names
   * @param args arguments
   */
  constructor(type: string, names: string[], args?: Record<string, any>, customArgumentToString?: ArgumentsToStringFn) {
    this.validateIdentifier(type);
    names.forEach(name => {
      this.validateIdentifier(name);
    });

    this.blockType = type;
    this.blockNames = names;
    this.arguments = args ? args : {};
    this.customArgumentToString = customArgumentToString;
  }

  /**
   * Get arguments.
   */
  getArguments(): Record<string, any> {
    return this.arguments;
  }

  /**
   * Get argument by key.
   * 
   * @param key key
   */
  getArgument(key: string): any {
    return this.arguments[key];
  }

  /**
   * Set argument.
   * 
   * @param key key
   * @param value value
   */
  setArgument(key: string, value: any): Block {
    this.arguments[key] = value;
    return this;
  }

  /**
   * Set arguments.
   * 
   * @param args arguments
   */
  setArguments(args: Record<string, any>): Block {
    for (const key in args) {
      this.arguments[key] = args[key];
    }
    return this;
  }

  /**
   * Delete argument by key.
   * 
   * @param key key
   */
  deleteArgument(key: string): Block {
    delete this.arguments[key];
    return this;
  }

  /**
   * To Terraform representation.
   * 
   * @param version Terraform version
   */
  toTerraform(version: TerraformVersion): string {
    let str = this.blockType;
    this.blockNames.forEach(name => {
      str += ` "${name}"`;
    });
    str += '{\n';
    str += TerraformGeneratorUtils.argumentsToString(version, this.arguments, this.customArgumentToString);
    str += '}\n\n';
    return str;
  }

  /**
   * Represent block as argument.
   */
  abstract asArgument(): Argument;

  /**
   * Get block's attribute.
   * 
   * @param name attribute name
   */
  abstract attr(name: string): Attribute;

  /**
   * Block's id attribute.
   */
  get id(): Attribute {
    return this.attr('id');
  }

  private validateIdentifier(identifier: string): void {
    if (!identifier.match(/^[a-zA-Z_\-]{1}[0-9a-zA-Z_\-]*$/)) {
      throw new Error(`Invalid identifier: ${identifier}`);
    }
  }

}
