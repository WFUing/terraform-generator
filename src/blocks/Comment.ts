import { Argument, Attribute } from '../arguments';
import { Block } from '.';

/**
 * @category Block
 */
export class Comment extends Block<Record<string, never>> {

  readonly comment: string;

  /**
   * Construct comment.
   *
   * @param comment comment
   */
  constructor(comment: string) {
    super('comment', [], {});

    this.comment = comment;
  }

  override toTerraform(): string {
    return `${this.comment.trim().split('\n').map(line => line.trim() ? `# ${line.trim()}` : '').join('\n')}\n\n`;
  }

  override asArgument(): Argument {
    throw new Error('Inaccessible method.');
  }

  override attr(_name: string): Attribute {
    throw new Error('Inaccessible method.');
  }

}
