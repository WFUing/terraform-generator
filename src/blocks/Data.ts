import { Block, Argument, Attribute } from '..';

export default class Data extends Block {

  readonly type: string;
  readonly name: string;

  /**
   * Construct data source.
   * Refer to Terraform documentation on what can be put as type & arguments.
   * 
   * @param type type
   * @param name name
   * @param args arguments
   * @param innerBlocks inner blocks
   */
  constructor(type: string, name: string, args?: Record<string, any>, innerBlocks?: Block[]) {
    super('data', [type, name], args, innerBlocks);

    this.type = type;
    this.name = name;
  }

  asArgument(): Argument {
    return new Argument(`data.${this.type}.${this.name}`);
  }

  attr(name: string): Attribute {
    return new Attribute(this, name);
  }

}
