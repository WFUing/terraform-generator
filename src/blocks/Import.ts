import { Argument, Attribute } from '../arguments';
import { Block, Resource } from '.';

/**
 * @category Block
 */
export interface ImportArgs {
  to: Resource;
  id: string;
  provider?: Argument;
}

/**
 * @category Block
 */
export class Import extends Block<ImportArgs> {

  /**
   * Construct import.
   *
   * Refer to Terraform documentation on what can be put as arguments.
   *
   * @param args arguments
   */
  constructor(args: ImportArgs) {
    super('import', [], args);
  }

  override asArgument(): Argument {
    throw new Error('Inaccessible method.');
  }

  override attr(_name: string): Attribute {
    throw new Error('Inaccessible method.');
  }

}
