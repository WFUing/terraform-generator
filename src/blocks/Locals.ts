import { Argument, Attribute } from '../arguments';
import { BlockArgs } from '../utils';
import { Block } from '.';

/**
 * @category Block
 */
export class Locals extends Block {

  /**
   * Construct locals.
   *
   * Refer to Terraform documentation on what can be put as arguments.
   *
   * @param args arguments
   */
  constructor(args: BlockArgs) {
    super('locals', [], args);
  }

  override asArgument(): Argument {
    throw new Error('Inaccessible method.');
  }

  override attr(_name: string): Attribute {
    throw new Error('Inaccessible method.');
  }

  arg(name: string): Argument {
    return new Argument(`local.${name}`);
  }

}
