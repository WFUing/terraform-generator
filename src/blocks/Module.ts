import { Argument, Attribute } from '../arguments';
import { TerraformArgs } from '../utils';
import { Block } from '.';

/**
 * @category Block
 */
export interface ModuleArgs extends TerraformArgs {
  source: string;
  version?: string;
}

/**
 * @category Block
 */
export class Module extends Block<ModuleArgs> {

  readonly name: string;

  /**
   * Construct module.
   *
   * Refer to Terraform documentation on what can be put as arguments.
   *
   * @param name name
   * @param args arguments
   */
  constructor(name: string, args: ModuleArgs) {
    super('module', [name], args);

    this.name = name;
  }

  override asArgument(): Argument {
    return new Argument(`module.${this.name}`);
  }

  override attr(name: string): Attribute {
    return new Attribute(this, name);
  }

}
