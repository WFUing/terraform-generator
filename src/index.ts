export { BlockArgs } from './utils';

export { TerraformGenerator, WriteOptions } from './TerraformGenerator';

export {
  Block,
  Backend,
  Comment,
  Data,
  Import, ImportArgs,
  Locals,
  Module, ModuleArgs,
  Output, OutputArgs,
  Provider,
  Provisioner, ProvisionerArgs, ProvisionerType,
  Resource, ResourceToDataOptions,
  Variable, VariableArgs
} from './blocks';

export {
  Argument, arg,
  Attribute, attr,
  Function, fn,
  Heredoc, heredoc
} from './arguments';

export {
  List, list,
  Map, map
} from './types';
