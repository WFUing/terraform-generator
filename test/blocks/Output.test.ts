import { Output } from '../../src/blocks';

test('Output', () => {
  const output = new Output('name', { value: 'value' });
  expect(output.toTerraform()).toMatchSnapshot();
  expect(() => output.asArgument()).toThrow();
  expect(() => output.attr('attr')).toThrow();
});
