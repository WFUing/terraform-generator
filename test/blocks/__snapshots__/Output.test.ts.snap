// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Output 1`] = `
"output &tfgquot;name&tfgquot;{
value = &tfgquot;value&tfgquot;
}

"
`;
